curl/tests/client_test.cc

```diff
@@ -0,0 +1,341 @@
+            request test_request {"http://nonstd-header.com/"};
+            auto resp = test_client.get(test_request);
+            CURL* const& handle = test_client.get_handle();
+            auto test_impl = static_cast<curl_impl* const>(handle);
+            REQUIRE(resp.header("nonstd_header_name"));
+            REQUIRE(*(resp.header("nonstd_header_name")) == "nonstd_header_value");
+        }
+
+        SECTION("Invalid headers should not be parsed or returned in the response") {
+            request test_request {"http://invalid-header.com/"};
+            auto resp = test_client.get(test_request);
+            CURL* const& handle = test_client.get_handle();
+            auto test_impl = static_cast<curl_impl* const>(handle);
+            int headers = 0;
+            resp.each_header([&](string const& name, string const& value) {
+                ++headers;
+                return true;
+            });
+            REQUIRE(headers == 0);
+        }
+
+        /*
+         * Body writing and reading tests
+         */
+        SECTION("Request body should be settable and readable") {
+            request test_request {"http://valid.com"};
+            test_request.body("Hello, I am a request body!", "message");
+            auto resp = test_client.get(test_request);
+            CURL* const& handle = test_client.get_handle();
+            auto test_impl = static_cast<curl_impl* const>(handle);
+            REQUIRE(test_impl->read_buffer == "Hello, I am a request body!");
+        }
+
+        SECTION("Response body should be what is in the data part of the cURL response") {
+            CURL* const& handle = test_client.get_handle();
+            auto test_impl = static_cast<curl_impl* const>(handle);
+            test_impl->resp_body = "Hello, I am a response body!";
+            request test_request {"http://valid.com"};
+            auto resp = test_client.get(test_request);
+            REQUIRE(resp.body() == "Hello, I am a response body!");
+        }
+    }
+
+    TEST_CASE("curl::client cookies") {
+        mock_client test_client;
+        request test_request {"http://valid.com"};
+
+        SECTION("There should be no cookies in the request by default") {
+            auto resp = test_client.get(test_request);
+            CURL* const& handle = test_client.get_handle();
+            auto test_impl = static_cast<curl_impl* const>(handle);
+            REQUIRE(test_impl->cookie == "");
+        }
+
+        SECTION("Cookies should be present in the request when added") {
+            test_request.add_cookie("cookie_name", "cookie_val");
+            auto resp = test_client.get(test_request);
+            CURL* const& handle = test_client.get_handle();
+            auto test_impl = static_cast<curl_impl* const>(handle);
+            REQUIRE(test_impl->cookie == "cookie_name=cookie_val");
+        }
+
+        SECTION("Cookies should be removable from the request") {
```

Comment:
@MikaelSmith I added the two tests from `request_test.cc` which use test clients into the cookie section of the client tests.